README
NAME
    POE::Component::IRC::Plugin::YouAreDoingItWrong - show people what they
    are doing wrong by giving links to http://doingitwrong.com images.

SYNOPSIS
        use strict;
        use warnings;

        use POE qw(Component::IRC Component::IRC::Plugin::YouAreDoingItWrong);

        my $irc = POE::Component::IRC->spawn(
                nick    => 'WrongBot',
                server  => 'irc.freenode.net',
                port    => 6667,
                ircname => 'You Are Doing It Wrong Bot',
        ) or die "Oh noes :( $!";

        POE::Session->create(
            package_states => [
                main => [ qw( _start  irc_001 ) ],
            ],
        );

        $poe_kernel->run();

        sub _start {
            $irc->yield( register => 'all' );

            # register our plugin
            $irc->plugin_add(
                'Wrong' => POE::Component::IRC::Plugin::YouAreDoingItWrong->new
            );

            $irc->yield( connect => { } );
            undef;
        }

        sub irc_001 {
            my ( $kernel, $sender ) = @_[ KERNEL, SENDER ];
            $kernel->post( $sender => join => '#zofbot' );

            undef;
        }

        --

        [13:00:27] <Zoffix> WrongBot, doing it wrong Zoffix
        [13:00:27] <WrongBot> Zoffix, you are doing it wrong: http://www.doingitwrong.com/wrong/20070527-113353.jpg
        [13:00:41] <Zoffix> WrongBot, doing it wrong
        [13:00:42] <WrongBot> You are doing it wrong: http://www.doingitwrong.com/wrong/1487_kolo.jpg

DESCRIPTION
    The module is a POE::Component::IRC plugin which, when triggered,
    fetches links to images from <http://doingitwrong.com> and posts them
    into the channel, optionally addressing a specific person.

CONSTRUCTOR
  new
        $irc->plugin_add(
            'Wrong' => POE::Component::IRC::Plugin::YouAreDoingItWrong->new
        );

        $irc->plugin_add(
            'Wrong' => POE::Component::IRC::Plugin::YouAreDoingItWrong->new(
                auto           => 0,                 # do not auto respond
                trigger        => qr/^wrong/i,       # trigger
                banned         => [ qr/aol\.com$/ ], # ignore AOL users
                response_event => 'diw_event',       # event to send response to
                debug          => 1,                 # enable some debug output
            )
        );

    Returns a POE::Component::IRC::Plugin object suitable to be fed to
    POE::Component::IRC "plugin_add()" method. Takes a few arguments, *all
    of which are optional*. The arguments may be as follows:

   auto
        ->new( auto => 1 );

    Optional. The "auto" argument specifies whether or not the plugin should
    auto respond with "privmsg" to the channel when triggered. If "auto" is
    set to a true value the plugin will automatically respond when
    triggered, when "auto" is set to a false value the plugin will not send
    any messages to the server and you'll have to listen to the event emited
    by the plugin (see below). In other words, if you are unhappy with the
    default behaviour you may turn "auto" off (by setting it to a false
    value) and set up a handler on the plugin event to do exactly what you
    want. Defaults to: 1

   trigger
        ->new( trigger => qr/^doing it wrong/i );

    Optional. Takes a regex as an argument which specifies the trigger for
    the plugin. Defaults to: "qr/^doing it wrong/i" Note: plugin responds to
    *addressed* messages, In other words, if you set the trigger to
    "qr/^diw/i" and your bot's nick name is "WrongBot" the plugin will be
    triggered by saying "WrongBot, diw". Note 2: if after removing bot's
    nickname and the trigger the left over will match "m/(\S+)\s*$/" the
    capture will be prepended to the output, this is so you could address a
    specific person:

        [13:00:27] <Zoffix> WrongBot, doing it wrong Zoffix
        [13:00:27] <WrongBot> Zoffix, you are doing it wrong: http://www.doingitwrong.com/wrong/20070527-113353.jpg
        [13:00:41] <Zoffix> WrongBot, doing it wrong
        [13:00:42] <WrongBot> You are doing it wrong: http://www.doingitwrong.com/wrong/1487_kolo.jpg

   banned
        ->new( banned => [ qr/aol\.com$/ ] );

    Optional. Takes an arrayref of regexes as an argument. If the user mask
    of the person who triggered the plugin matches any of the regexes in the
    "banned" arrayref the plugin will ignore that person. Defaults to: "[]"
    (no bans are set)

   response_event
        ->new( response_event => 'diw_event' );

    Optional. Specifies the name of the event to emit when plugin is
    triggered. See EMITED EVENTS section for details. Defaults to:
    "irc_you_are_doing_it_wrong_response"

   debug
        ->new( debug => 1 );

    Optional. When set to a true value plugin will print out a bit of debug
    information. Defaults to: 0 (no debug info is printed out)

   eat
        ->new( eat => 0 );

    If set to a false value plugin will return a "PCI_EAT_NONE" after
    responding. If eat is set to a true value, plugin will return a
    "PCI_EAT_ALL" after responding. See POE::Component::IRC::Plugin
    documentation for more information if you are interested. Defaults to: 1

EMITED EVENTS
  response_event
        $VAR1 = {
            'what' => 'WrongBot, doing it wrong',
            'who' => 'Zoffix!n=Zoffix@unaffiliated/zoffix',
            'channel' => '#zofbot',
            'nick_to_address' => undef,
            'error' => undef,
            'pic' => bless( do{\(my $o = 'http://www.doingitwrong.com/wrong/20070929-012129.jpg')}, 'URI::http' )
        };

    The event handler set up to handle the event, name of which you can
    specify in the "response_event" of the construcor (it defaults to
    "irc_you_are_doing_it_wrong_response") will recieve input from the
    plugin every time it's triggered. The input will come in the form of a
    hashref in "ARG0". The keys of that hashref are as follows:

   what
        { 'what' => 'WrongBot, doing it wrong' }

    The "what" key will contain the message which triggered the plugin.

   who
        { 'who' => 'Zoffix!n=Zoffix@unaffiliated/zoffix' }

    The "who" key will contain the mask of the user who triggered the
    plugin.

   nick_to_address
        { 'nick_to_address' => undef }

    The "nick_to_address" key will contain the nick of the person to address
    if the plugin was triggered in the "address someone" mode (see
    description of the "trigger" argument to the contructor) or "undef" if
    plugin was triggered in normal mode.

   error
        { error => undef }

    If an error occured during while fetching the link to a random image on
    <http://doingitwrong.com> the "error" key will contain the error
    message, otherwise it will be "undef".

   pic
        { 'pic' => bless( do{\(my $o = 'http://www.doingitwrong.com/wrong/20070929-012129.jpg')}, 'URI::http' ) }

    If no errors occured (see "error" above), the "pic" key will contain a
    URI object pointing to one of the images on
    <http://www.doingitwrong.com>.

AUTHOR
    Zoffix Znet, "<zoffix at cpan.org>" (<http://zoffix.com>,
    <http://haslayout.net>)

BUGS
    Please report any bugs or feature requests to
    "bug-poe-component-irc-plugin-youaredoingitwrong at rt.cpan.org", or
    through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Component-IRC-Plugin
    -YouAreDoingItWrong>. I will be notified, and then you'll automatically
    be notified of progress on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc POE::Component::IRC::Plugin::YouAreDoingItWrong

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-IRC-Plugin-Y
        ouAreDoingItWrong>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/POE-Component-IRC-Plugin-YouAreDoingItWron
        g>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/POE-Component-IRC-Plugin-YouAreDoingI
        tWrong>

    *   Search CPAN

        <http://search.cpan.org/dist/POE-Component-IRC-Plugin-YouAreDoingItW
        rong>

COPYRIGHT & LICENSE
    Copyright 2008 Zoffix Znet, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for POE-Component-IRC-Plugin-YouAreDoingItWrong

0.0102  2013-12-16
        - Fixed a fatal bug due to disappeared _send_event in PoCo::IRC
        - Minor fixes to comply with the specs and Kwalitee game
        - Added autogeneration of README from POD
        - Moved author tests to xt/

0.02    2008-02-12
        - fixed documentation discrepancy for
          constructors 'eat' argument
          as well as plugin's return value (PCI_EAT_* )

0.01    2008-02-10
        - First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'POE::Component::IRC::Plugin::YouAreDoingItWrong',
    license             => 'perl',
    dist_author         => 'Zoffix Znet <zoffix@cpan.org>',
    dist_version_from   => 'lib/POE/Component/IRC/Plugin/YouAreDoingItWrong.pm',
    build_requires => {
        'Test::More' => 0,
    },
    requires => {
        'POE'                                             => 0.9999,
        'POE::Component::WWW::DoingItWrongCom::RandImage' => 0.01,
        'POE::Component::IRC::Plugin'                     => 0.09,
        'perl' => '5.006',
    },
    configure_requires => { 'Module::Build' => 0 },
    add_to_cleanup      => [ 'POE-Component-IRC-Plugin-YouAreDoingItWrong-*' ],
    create_makefile_pl => 'traditional',
);
if ( $ENV{RELEASE_TESTING} ) {
    system('pod2text lib/POE/Component/IRC/Plugin/YouAreDoingItWrong.pm > README');
}
$builder->create_build_script();


examples/wrong_bot.pl
#!/usr/bin/env perl

use strict;
use warnings;

use lib qw(lib ../lib);
use POE qw(Component::IRC Component::IRC::Plugin::YouAreDoingItWrong);

my @Channels = ( '#zofbot' );

my $irc = POE::Component::IRC->spawn(
        nick    => 'WrongBot',
        server  => 'irc.freenode.net',
        port    => 6667,
        ircname => 'You Are Doing It Wrong Bot',
) or die "Oh noes :( $!";

POE::Session->create(
    package_states => [
        main => [
            qw(
                _start
                irc_001
                _default
                irc_you_are_doing_it_wrong_response
            )
        ],
    ],
);

$poe_kernel->run();

sub irc_you_are_doing_it_wrong_response {

    my $in = $_[ARG0];
    print "\n";
    foreach my $data ( qw(what who channel pic error) ) {
        next
            unless defined $in->{ $data };
        print "    $data => $in->{ $data }\n";
    }
    print "\n";
}

sub _start {
    $irc->yield( register => 'all' );

    # register our plugin
    $irc->plugin_add(
        'Wrong' => POE::Component::IRC::Plugin::YouAreDoingItWrong->new(
            debug => 1,
        )
    );

    $irc->yield( connect => { } );
    undef;
}

sub irc_001 {
    my ( $kernel, $sender ) = @_[ KERNEL, SENDER ];
    $kernel->post( $sender => join => $_ )
        for @Channels;

    undef;
}


sub _default {
    my ($event, $args) = @_[ARG0 .. $#_];

    my @output = ( "$event: " );

    foreach my $arg ( @$args ) {
        if ( ref($arg) eq 'ARRAY' ) {
                push( @output, "[" . join(" ,", @$arg ) . "]" );
        } else {
                push ( @output, "'$arg'" );
        }
    }
    print STDOUT join ' ', @output, "\n";
    return 0;
}

lib/POE/Component/IRC/Plugin/YouAreDoingItWrong.pm
package POE::Component::IRC::Plugin::YouAreDoingItWrong;

use warnings;
use strict;

our $VERSION = '0.03';

use POE qw(Component::WWW::DoingItWrongCom::RandImage);
use POE::Component::IRC::Plugin qw(:ALL);

sub new {
    my $package = shift;
    my %args = @_;
    $args{ lc $_ } = delete $args{ $_ } for keys %args;

    # fill in the defaults
    %args = (
        debug          => 0,
        auto           => 1,
        trigger        => qr/^doing it wrong/i,
        banned         => [],
        response_event => 'irc_you_are_doing_it_wrong_response',

        %args,
    );

    return bless \%args, $package;
}

sub _start {
    my ( $kernel, $self ) = @_[ KERNEL, OBJECT ];
    $self->{_session_id} = $_[SESSION]->ID();
    $kernel->refcount_increment( $self->{_session_id}, __PACKAGE__ );

    $self->{poco} = POE::Component::WWW::DoingItWrongCom::RandImage->spawn(
        debug => $self->{debug},
    );

    undef;
}

sub PCI_register {
    my ( $self, $irc ) = splice @_, 0, 2;

    $self->{irc} = $irc;

    $irc->plugin_register( $self, 'SERVER', qw(public) );

    $self->{_session_id} = POE::Session->create(
        object_states => [
            $self => [
                qw(
                    _start
                    _shutdown
                    _pic
                    _get_pic
                )
            ],
        ],
    )->ID;

    return 1;
}

sub _shutdown {
    my ($kernel, $self) = @_[ KERNEL, OBJECT ];
    $self->{poco}->shutdown;
    $kernel->alarm_remove_all();
    $kernel->refcount_decrement( $self->{_session_id}, __PACKAGE__ );
    undef;
}

sub PCI_unregister {
    my $self = shift;

    # Plugin is dying make sure our POE session does as well.
    $poe_kernel->call( $self->{_session_id} => '_shutdown' );

    delete $self->{irc};

    return 1;
}

sub S_public {
    my ( $self, $irc ) = splice @_, 0, 2;
    my $who     = ${ $_[0] };
    my $channel = ${ $_[1] }->[0];
    my $message = ${ $_[2] };

    foreach my $ban_re ( @{ $self->{banned} || [] } ) {
        return PCI_EAT_NONE
            if $who =~ /$ban_re/;
    }

    my $my_nick = $irc->nick_name();
    my ($what) = $message =~ m/^\s*\Q$my_nick\E[\:\,\;\.]?\s*(.*)$/i;

    return PCI_EAT_NONE
         if !defined $what or !($what =~ s/$self->{trigger}//);

    my ( $nick_to_address ) = $what =~ /(\S+)\s*$/;

    warn "Got PUBLIC input: [ who => $who, channel => $channel, "
            . "what => $what ]"
        if $self->{debug};
    $poe_kernel->post( $self->{_session_id} => _get_pic => {
            _what            => $message,
            _who             => $who,
            _channel         => $channel,
            _nick_to_address => $nick_to_address,
        }
    );

    return $self->{eat} ? PCI_EAT_ALL : PCI_EAT_NONE;
}

sub _get_pic {
    $_[OBJECT]->{poco}->fetch( {
            event            => '_pic',
            %{ $_[ARG0] },
        },
    );
}

sub _pic {
    my ( $kernel, $self, $result ) = @_[ KERNEL, OBJECT, ARG0 ];
    my $message;
    if ( $result->{error} ) {
        $message = "Error: $result->{error}";
    }
    else {
        $message = "You are doing it wrong: $result->{out}";
    }

    $self->{irc}->send_event(
        $self->{response_event} => {
            pic   => $result->{out},
            error => $result->{error},
            map { $_ => $result->{"_$_"} }
                qw(who what channel nick_to_address)
        }
    );

    if ( $self->{auto} ) {
        if ( defined $result->{_nick_to_address} ) {
            $message = "$result->{_nick_to_address}, \l$message";
        }

        $kernel->post(
            $self->{irc} => privmsg => $result->{_channel} => $message
        );
    }

    undef;
}

1;

__END__

=encoding utf8

=head1 NAME

POE::Component::IRC::Plugin::YouAreDoingItWrong - show people what they
are doing wrong by giving links to http://doingitwrong.com images.

=head1 SYNOPSIS

    use strict;
    use warnings;

    use POE qw(Component::IRC Component::IRC::Plugin::YouAreDoingItWrong);

    my $irc = POE::Component::IRC->spawn(
            nick    => 'WrongBot',
            server  => 'irc.freenode.net',
            port    => 6667,
            ircname => 'You Are Doing It Wrong Bot',
    ) or die "Oh noes :( $!";

    POE::Session->create(
        package_states => [
            main => [ qw( _start  irc_001 ) ],
        ],
    );

    $poe_kernel->run();

    sub _start {
        $irc->yield( register => 'all' );

        # register our plugin
        $irc->plugin_add(
            'Wrong' => POE::Component::IRC::Plugin::YouAreDoingItWrong->new
        );

        $irc->yield( connect => { } );
        undef;
    }

    sub irc_001 {
        my ( $kernel, $sender ) = @_[ KERNEL, SENDER ];
        $kernel->post( $sender => join => '#zofbot' );

        undef;
    }

    --

    [13:00:27] <Zoffix> WrongBot, doing it wrong Zoffix
    [13:00:27] <WrongBot> Zoffix, you are doing it wrong: http://www.doingitwrong.com/wrong/20070527-113353.jpg
    [13:00:41] <Zoffix> WrongBot, doing it wrong
    [13:00:42] <WrongBot> You are doing it wrong: http://www.doingitwrong.com/wrong/1487_kolo.jpg

=head1 DESCRIPTION

The module is a L<POE::Component::IRC> plugin which, when triggered,
fetches links to images from L<http://doingitwrong.com> and posts them
into the channel, optionally addressing a specific person.

=head1 CONSTRUCTOR

=head2 new

    $irc->plugin_add(
        'Wrong' => POE::Component::IRC::Plugin::YouAreDoingItWrong->new
    );

    $irc->plugin_add(
        'Wrong' => POE::Component::IRC::Plugin::YouAreDoingItWrong->new(
            auto           => 0,                 # do not auto respond
            trigger        => qr/^wrong/i,       # trigger
            banned         => [ qr/aol\.com$/ ], # ignore AOL users
            response_event => 'diw_event',       # event to send response to
            debug          => 1,                 # enable some debug output
        )
    );

Returns a L<POE::Component::IRC::Plugin> object suitable to be fed to
L<POE::Component::IRC> C<plugin_add()> method. Takes a few arguments,
I<all of which are optional>. The arguments may be as follows:

=head3 auto

    ->new( auto => 1 );

B<Optional>. The C<auto> argument specifies whether or not the plugin
should auto respond with C<privmsg> to the channel when triggered. If
C<auto> is set to a true value the plugin will automatically respond
when triggered, when C<auto> is set to a false value the plugin will
not send any messages to the server and you'll have to listen to the
event emited by the plugin (see below). In
other words, if you are unhappy with the default behaviour you may
turn C<auto> off (by setting it to a false value) and set up a handler
on the plugin event to do exactly what you want. B<Defaults to:> C<1>


=head3 trigger

    ->new( trigger => qr/^doing it wrong/i );

B<Optional>. Takes a regex as an argument which specifies the
trigger for the plugin. B<Defaults to:> C<qr/^doing it wrong/i>
B<Note:> plugin responds to I<addressed> messages,
In other words, if you set the trigger to C<qr/^diw/i> and your
bot's nick name is C<WrongBot> the plugin will be triggered by saying
C<WrongBot, diw>. B<Note 2:> if after removing bot's nickname and the
trigger the left over will match C<m/(\S+)\s*$/> the capture will be
prepended to the output, this is so you could address a specific person:

    [13:00:27] <Zoffix> WrongBot, doing it wrong Zoffix
    [13:00:27] <WrongBot> Zoffix, you are doing it wrong: http://www.doingitwrong.com/wrong/20070527-113353.jpg
    [13:00:41] <Zoffix> WrongBot, doing it wrong
    [13:00:42] <WrongBot> You are doing it wrong: http://www.doingitwrong.com/wrong/1487_kolo.jpg

=head3 banned

    ->new( banned => [ qr/aol\.com$/ ] );

B<Optional>. Takes an arrayref of regexes as an argument. If the user mask
of the person who triggered the plugin matches any of the regexes in
the C<banned> arrayref the plugin will ignore that person. B<Defaults to:>
C<[]> (no bans are set)

=head3 response_event

    ->new( response_event => 'diw_event' );

B<Optional>. Specifies the name of the event to emit when plugin is
triggered. See EMITED EVENTS section for details. B<Defaults to:>
C<irc_you_are_doing_it_wrong_response>

=head3 debug

    ->new( debug => 1 );

B<Optional>. When set to a true value plugin will print out a bit of
debug information. B<Defaults to:> C<0> (no debug info is printed out)

=head3 eat

    ->new( eat => 0 );

If set to a false value plugin will return a C<PCI_EAT_NONE> after
responding. If eat is set to a true value, plugin will return a
C<PCI_EAT_ALL> after responding. See L<POE::Component::IRC::Plugin>
documentation for more information if you are interested. B<Defaults to:>
C<1>

=head1 EMITED EVENTS

=head2 response_event

    $VAR1 = {
        'what' => 'WrongBot, doing it wrong',
        'who' => 'Zoffix!n=Zoffix@unaffiliated/zoffix',
        'channel' => '#zofbot',
        'nick_to_address' => undef,
        'error' => undef,
        'pic' => bless( do{\(my $o = 'http://www.doingitwrong.com/wrong/20070929-012129.jpg')}, 'URI::http' )
    };

The event handler set up to handle the event, name of which you
can specify in the C<response_event> of the construcor (it defaults to
C<irc_you_are_doing_it_wrong_response>) will recieve input from the
plugin every time it's triggered. The input will come in the form
of a hashref in C<ARG0>. The keys of that hashref are as follows:

=head3 what

    { 'what' => 'WrongBot, doing it wrong' }

The C<what> key will contain the message which triggered the plugin.

=head3 who

    { 'who' => 'Zoffix!n=Zoffix@unaffiliated/zoffix' }

The C<who> key will contain the mask of the user who triggered the plugin.

=head3 nick_to_address

    { 'nick_to_address' => undef }

The C<nick_to_address> key will contain the nick of the person to address
if the plugin was triggered in the "address someone" mode
(see description of the C<trigger> argument to the contructor) or C<undef>
if plugin was triggered in normal mode.

=head3 error

    { error => undef }

If an error occured during while fetching the link to a random image
on L<http://doingitwrong.com> the C<error> key will contain the error
message, otherwise it will be C<undef>.

=head3 pic

    { 'pic' => bless( do{\(my $o = 'http://www.doingitwrong.com/wrong/20070929-012129.jpg')}, 'URI::http' ) }

If no errors occured (see C<error> above), the C<pic> key will contain
a L<URI> object pointing to one of the images on
L<http://www.doingitwrong.com>.

=head1 AUTHOR

Zoffix Znet, C<< <zoffix at cpan.org> >>
(L<http://zoffix.com>, L<http://haslayout.net>)

=head1 BUGS

Please report any bugs or feature requests to C<bug-poe-component-irc-plugin-youaredoingitwrong at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Component-IRC-Plugin-YouAreDoingItWrong>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc POE::Component::IRC::Plugin::YouAreDoingItWrong

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-IRC-Plugin-YouAreDoingItWrong>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/POE-Component-IRC-Plugin-YouAreDoingItWrong>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/POE-Component-IRC-Plugin-YouAreDoingItWrong>

=item * Search CPAN

L<http://search.cpan.org/dist/POE-Component-IRC-Plugin-YouAreDoingItWrong>

=back

=head1 COPYRIGHT & LICENSE

Copyright 2008 Zoffix Znet, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

